attestation/github/github.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2021 The Witness Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/testifysec/go-witness/attestation"
+	"github.com/testifysec/go-witness/attestation/jwt"
+	"github.com/testifysec/go-witness/cryptoutil"
+)
+
+const (
+	Name    = "github"
+	Type    = "https://witness.dev/attestations/github/v0.1"
+	RunType = attestation.PreRunType
+)
+
+const (
+	tokenAudience = "witness"
+	jwksURL       = "https://token.actions.githubusercontent.com/.well-known/jwks"
+)
+
+// This is a hacky way to create a compile time error in case the attestor
+// doesn't implement the expected interfaces.
+var (
+	_ attestation.Attestor   = &Attestor{}
+	_ attestation.Subjecter  = &Attestor{}
+	_ attestation.BackReffer = &Attestor{}
+)
+
+func init() {
+	attestation.RegisterAttestation(Name, Type, RunType, func() attestation.Attestor {
+		return New()
+	})
+}
+
+type ErrNotGitlab struct{}
+
+func (e ErrNotGitlab) Error() string {
+	return "not in a github ci job"
+}
+
+type Attestor struct {
+	JWT          *jwt.Attestor `json:"jwt,omitempty"`
+	CIConfigPath string        `json:"ciconfigpath"`
+	PipelineID   string        `json:"pipelineid"`
+	PipelineName string        `json:"pipelinename"`
+	PipelineUrl  string        `json:"pipelineurl"`
+	ProjectUrl   string        `json:"projecturl"`
+	RunnerID     string        `json:"runnerid"`
+	CIHost       string        `json:"cihost"`
+	CIServerUrl  string        `json:"ciserverurl"`
+	RunnerArch   string        `json:"runnerarch"`
+	RunnerOS     string        `json:"runneros"`
+	jwksURL      string        `json:"jwksurl"`
+	tokenURL     string        `json:"tokenurl"`
```

Comment:
This and the other two fields linting is yelling about will never appear in the marshaled json because they are unexported.  They should either be exported or not have json tags if they're not meant to be in the marashaled json.